### PR TITLE
Add support for NUnit version 4.

### LINKED ISSUE #562: Warn use of params for assertion messages
Ref issue https://github.com/nunit/nunit/issues/4413 and Issue https://github.com/nunit/nunit/issues/3936 - introduction of CallerArgumentExpressions (CAE) and https://github.com/nunit/nunit/issues/4415.

Since we need to have the CAE at the end of the parameter list, we can't have the params there. The use of params for messages are now superseded by the use of string interpolation.

The analyzer should warn about this usage and suggest conversion to an interpolated string.

### LINKED ISSUE #606: Support for NUnit 4 legacy asserts
In NUnit 4 the classic `Assert` statements will be changed to `ClassicAssert`, and will be in the namespace `NUnit.Framework.Legacy`.

The rules that allow converting the classic asserts to the constraint model should include the `ClassicAssert` class.

Examples of using the classic asserts in NUnit 4 can be found here:  https://github.com/nunit/nunit-csharp-samples

### LINKED ISSUE #610: Ensure Test Code works with NUnit4
Upgrade test project to use NUnit4 and make necessary changes.

### LINKED ISSUE #617: Update .Within makes no sense rule
Because of the newly added [`PropertyComparer`](https://github.com/nunit/docs/issues/812) the `Within` rule is no longer correct. If a class does not implement `IEquatable`, we now pass each property to the comparer, passing along the tolerance.

The rule needs to stay in place for types implementing `IEquatable`.

### LINKED ISSUE #618: Make Classic Conversion rule for StringAssert
NUnit4 build failed on a `StringAssert` which didn't trigger a NUnit.Analyzer conversion rule.

### LINKED ISSUE #620: Make Classic Conversion rule for CollectionAssert

### PR DESCRIPTION
**Pull request by manfred-brands**
Detect ClassicAssert

Fixes #610 
Fixes #606
Fixes #562
Fixes #617
Fixes #618 
Fixes #620

### CODE REVIEW

**Comment by manfred-brands** [comment]
@mikkelbu I have tested this extensively with my work projects and converted about 50,000 classic asserts, including StringAssert and CollectionAssert. In addition it converted string format-specification+params into formattable strings.

I didn't make separate PRs as some build on others (params -> interpolatable).

We can release this as 3.90.
For 4.0 we then drop the netstandard1.6 build.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs [hunk 1]
```diff
@@ -36,10 +43,10 @@ private class B
...
 
         [TestCase("")]
-        [TestCase("Assert.NotNull(string.Empty)")]
-        [TestCase("Assert.IsNull(s)")]
-        [TestCase("Assert.Null(s)")]
-        [TestCase("Assert.That(s, Is.Null)")]
+        [TestCase("ClassicAssert.NotNull(string.Empty)")]
+        [TestCase("ClassicAssert.IsNull(s)")]
+        [TestCase("ClassicAssert.Null(s)")]
+        [TestCase("ClassicAssert.That(s, Is.Null)")]
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:49]: Should these 4 occurrences also use `$"{ClassicAssert}` instead of `ClassicAssert`

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs [hunk 2]
```diff
@@ -56,8 +63,8 @@ public void Test(string? s)
                 testCode);
         }
 
-        [TestCase("Assert.NotNull(s)")]
-        [TestCase("Assert.IsNotNull(s)")]
+        [TestCase("ClassicAssert.NotNull(s)")]
+        [TestCase("ClassicAssert.IsNotNull(s)")]
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:67]: Same comment as above

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs [hunk 3]
```diff
@@ -75,8 +82,8 @@ public void Test(string? s)
...
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:86]: Same comment as above

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs [hunk 4]
```diff
@@ -109,7 +116,7 @@ public void Test(string? s)
 
                 private static string DoSomething(string? s)
                 {
-                    Assert.NotNull(s);
+                    ClassicAssert.NotNull(s);
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:119]: Same comment as above

**Review by mikkelbu — COMMENTED** [review]
Great work @manfred-brands 👍 . I've reviewed the majority of the changes, but I still need to review 21 files (mostly the new analyzers and code-fixes). I should have time for this tomorrow.

So far I've only some minor comments and questions.

File: src/nunit.analyzers.tests/nunit.analyzers.tests.csproj [hunk 5]
```diff
@@ -1,7 +1,12 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <RootNamespace>NUnit.Analyzers.Tests</RootNamespace>
-    <TargetFrameworks>net7.0;net462</TargetFrameworks>
+    <TargetFrameworks>net6.0;net462</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers.tests/nunit.analyzers.tests.csproj:4]: At one point I had a version conflict between a package pulled in by nunit and this.
Downgrading to the same version as the NUnit binaries solved it at that time.
But I cannot reproduce it now, maybe the dependency wasn't needed.

File: src/nunit.analyzers.tests/TestUtility.cs [hunk 6]
```diff
@@ -10,6 +10,12 @@ internal static class TestUtility
 using System;
 using System.Threading.Tasks;
 using NUnit.Framework;
+#if NUNIT4
+using NUnit.Framework.Legacy;
+#else
+using ClassicAssert = NUnit.Framework.Assert;
+#endif
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers.tests/TestUtility.cs:17]: Yes. The tests run against both nunit3 and nunit4. 
The code is updated to use the nunit4 `ClassicAssert`.
So when building the tests against NUnit3 we need an alias to fall back to `Assert` 
and when building against NUnit3 we need to include the new `Legacy` namespace.

File: src/nunit.analyzers.tests/WithinUsage/WithinUsageCodeFixTests.cs [hunk 7]
```diff
@@ -1,4 +1,8 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
 using Gu.Roslyn.Asserts;
+using Microsoft.CodeAnalysis;
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers.tests/WithinUsage/WithinUsageCodeFixTests.cs:5]: No, but the light grey is hard to see and not showing up in the editor sidebar. I have increased the level of IDE0005 to _warning_ so they show at least up in the editor and added a commit removing all unnecessary using statements.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs (hunk 1 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:49]: Thanks for noticing.

The DiagnosticSuppressors work on the Syntax level (which is text) and don't know about the `using ClassicAssert = NUnit.Framework.Assert;` alias. Initially I changed some cases, but then decided to support both nunit versions in the analyzer:

```
private static bool IsValidatedNotNullByExpression(string possibleNullReference, ExpressionSyntax expression)
{
    // Check if this is an Assert for the same symbol
    if (AssertHelper.IsAssert(expression, out string member, out ArgumentListSyntax? argumentList) ||
        AssertHelper.IsClassicAssert(expression, out member, out argumentList))
    {
```

Al other code refers to this as `ClassicAssert`, so I will remove the constant.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs (hunk 2 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:67]: Same answer as well.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs (hunk 4 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:119]: Using all ClassicAssert.

**Comment by manfred-brands** [comment]
I have added the two mentioned changes and rebased on latest master.

File: src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs (hunk 3 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DiagnosticSuppressors/DereferencePossiblyNullReferenceSuppressorTests.cs:76]: No

File: src/nunit.analyzers.tests/DocumentationTests.cs [hunk 8]
```diff
@@ -18,6 +18,9 @@ namespace NUnit.Analyzers.Tests
 {
     public class DocumentationTests
     {
+        // Overload of Append(InterpolatedStringHanlder) with IFormatProvider not available in .NET 462
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers.tests/DocumentationTests.cs]: Done.

File: src/nunit.analyzers.tests/WithinUsage/WithinUsageAnalyzerTests.cs [hunk 9]
```diff
@@ -312,5 +325,119 @@ public void Test({type}? a, {type}? b)
...
+            string testCode = TestUtility.WrapClassInNamespaceAndAddUsing(@$"
+    [TestFixture]
+    public sealed class TestClass
+    {{
+        [Test]
+        public void TestMethod()
+        {{
+            var instance = new UserDefinedType(1, 1.1, ""1.1"");
+
+            Assert.That(new UserDefinedType(1, 1.2, ""1.1""), Is.EqualTo(instance).Within(0.1));
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/WithinUsage/WithinUsageAnalyzerTests.cs:376]:
```suggestion
            Assert.That(new UserDefinedType(1, 1.2, ""1.1""), Is.EqualTo(instance).↓Within(0.1));
```

File: src/nunit.analyzers.tests/WithinUsage/WithinUsageAnalyzerTests.cs [hunk 10]
```diff
@@ -312,5 +325,119 @@ public void Test({type}? a, {type}? b)
...
+            string testCode = TestUtility.WrapClassInNamespaceAndAddUsing(@"
+    [TestFixture]
+    public sealed class TestClass
+    {
+        [Test]
+        public void TestMethod()
+        {
+            var instance = new SomeRecord(1, 1.1, ""1.1"");
+
+            Assert.That(new SomeRecord(1, 1.2, ""1.1""), Is.EqualTo(instance).Within(0.1));
```
Comment (mikkelbu) [inline src/nunit.analyzers.tests/WithinUsage/WithinUsageAnalyzerTests.cs:422]:
```suggestion
            Assert.That(new SomeRecord(1, 1.2, ""1.1""), Is.EqualTo(instance).↓Within(0.1));
```

**Review by mikkelbu — APPROVED** [review]
Tremendous work @manfred-brands 👍. I found CodeFixHelper.cs a little hard to follow, so perhaps we need some more tests of this (we can also postpone this - just so we can get all the improvements out). 

I'm happy to get this in (and release it) when you are

File: src/nunit.analyzers.tests/nunit.analyzers.tests.csproj (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikkelbu) [reply src/nunit.analyzers.tests/nunit.analyzers.tests.csproj:4]: I think it is fine to use the LTS (at least for now)

File: src/nunit.analyzers.tests/UpdateStringFormatToInterpolatableString/UpdateStringFormatToInterpolatableStringCodeFixTests.cs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+                const int actual = 42;
+                ↓Assert.That(actual, Is.EqualTo(42), ""Expected actual value to be 42, but was: {0}"", actual);
+            ");
+            var fixedCode = TestUtility.WrapInTestMethod(@"
+                const int actual = 42;
+                Assert.That(actual, Is.EqualTo(42), $""Expected actual value to be 42, but was: {actual}"");
+            ");
+            RoslynAssert.CodeFix(analyzer, fix, expectedDiagnostic, code, fixedCode);
+        }
+
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers.tests/UpdateStringFormatToInterpolatableString/UpdateStringFormatToInterpolatableStringCodeFixTests.cs:67]: Updated test

File: src/nunit.analyzers/Helpers/Constraints.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace NUnit.Analyzers.Helpers
+{
+    internal sealed class Constraints
+    {
+        public Constraints(string staticClass, string? modifier, string? constraintMethod, string? property1 = default, string? property2 = null)
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers/Helpers/Constraints.cs:10]: No idea how why I they became different. Changed both to `null` as being the prevalent one use in the project.

File: src/nunit.analyzers.tests/WithinUsage/WithinUsageAnalyzerTests.cs (hunk 10 above)
Comment (manfred-brands) [reply src/nunit.analyzers.tests/WithinUsage/WithinUsageAnalyzerTests.cs:422]: Thanks. Also added ↓ at line 376 which is a similar test.

File: src/nunit.analyzers/Helpers/CodeFixHelper.cs [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+                        if (formatArgument.DescendantNodesAndSelf(x => true, false)
+                                          .OfType<ConditionalExpressionSyntax>()
+                                          .Any())
+                        {
+                            // Colon is not allowed in an Interpolation, wrap expression in parenthesis: (expession).
+                            formatArgument = SyntaxFactory.ParenthesizedExpression(formatArgument);
+                        }
+
+                        int argumentAlignmentSpecification = formatSpecification.IndexOf(',', argumentSpecification, argumentSpecificationEnd - argumentSpecification);
+                        int argumentFormatSpecification = formatSpecification.IndexOf(':', argumentSpecification, argumentSpecificationEnd - argumentSpecification);
```
[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/nunit.analyzers/Helpers/CodeFixHelper.cs:136]: They were indeed missing. Added a `TestFormatModifiers` and a `TestEscapedBrace` tests to `UpdateStringFormatToInterpolatableString`

**Comment by manfred-brands** [comment]
@mikkelbu There are some discussions about [renaming the ClassicAsserts](https://github.com/nunit/nunit/discussions/4437#discussioncomment-7411290) which would mean changing this.
If you want to release it now, it probably should be a beta package.
